File: js/applications/isPalindromo.ts
import { Dequeue } from "@lib/dequeue";

export function isPalindromo(word?: string): boolean {
    if (!word) return false;

    const deque = new Dequeue<string>();

    const lowerCaseWord = word.toLocaleLowerCase();

    let isEqual = true;
    let firstChar, lastChar;

    for (let i = 0; i < lowerCaseWord.length; i++) {
        deque.addBack(lowerCaseWord.charAt(i));
    }

    while (deque.size() > 1 && isEqual) {
        firstChar = deque.removeFront();
        lastChar = deque.removeBack();

        if (firstChar !== lastChar) {
            isEqual = false;
        }
    }

    return isEqual;
}

const palindromes = ["ana", "not", "madam"];

palindromes.forEach(word => {
    console.log({ word, isPalindromo: isPalindromo(word) });
});

File: lua/applications/is_palindromo.lua
require("lib/dequeue")

---@param word string
function Palindromo(word)
    if not word or #word == 0 then
        return false
    end

    local dequeue = Dequeue.new()
    for i = 1, string.len(word), 1 do
        local letter = string.sub(word, i, i)
        dequeue:add_back(letter)
    end

    local result = dequeue:to_string()

    return result == word
end

local word = "ana"
local result = Palindromo(word)

print("The word:", word, " is a palindromo:", result)
